### ISSUE TITLE
How to improve ray log experience

### ISSUE THREAD

**Issue by guoyuhong** [ray-project/ray#2713]
Recently, we are trying to improve the ray log experience to better support user debugging.

Imaging that a user write a print function in his remote functions to output some debugging information. His script fails for some reason. He wants to find the log, but he does not know which machine his task is running. There are tons of logs in a big cluster. Even though he manages to find out the running machine. It is hard to find out the log from many worker logs. 

We are trying to write programs to provide users with the target log. However, one worker will be assigned with tasks from different drivers. That is to say in one worker log file, there will be output from different task of different drivers or even different users. 

Is there a way to output the log according to driver ID or output the message with the driver ID?

**Comment by guoyuhong** [comment]
@pcmoritz @robertnishihara @raulchen @richardliaw @imzhenyu @eric-jj @mylinyuzhi

**Comment by guoyuhong** [comment]
It is very hard to add a timestamp or driver stamp to a user message. Is it possible to add messages before and after a task? For example, the log may look like the following message:
```
Driver(driver_id) starts the task(task_id)
User log for this task.
Driver(driver_id) finishes the task(task_id)
```
This way will make the log verbose. Even the task does not output anything. There still be messages.

**Comment by guoyuhong** [comment]
@mitar

**Comment by ericl** [comment]
I've always thought the right organization for user logs would be by actor and task name, for example:
```
actor_A_worker_1.log
actor_A_worker_2.log
func_a_worker_1.log
func_b_worker_1.log
func_b_worker_2.log
```

**Comment by guoyuhong** [comment]
Thanks, @ericl @mitar for providing your ideas for a good start. We can discuss more and consider the interface for more scenarios.

**Comment by guoyuhong** [comment]
@mitar I noticed that and I read through. The decorator and the context looks very good. I hope you can merge the code soon! Then we may reuse it in other scenario. ;-)

**Comment by mitar** [comment]
Hm, that code is not part of the Ray and I had no plans to integrate it with Ray. But if Ray people would like, we could move some parts like I describe above into Ray?

**Comment by ericl** [comment]
We could possibly write the logs to redis instead of MongoDB and surface errors / etc at the driver? I know a common complaint from users is that they have a hard time finding the logs from actors and tasks when running in cluster mode.

I guess one concern is space usage in redis, but it should be possible to set reasonable limits on usage for logs if they're just for debugging.

**Comment by pcmoritz** [comment]
Here is a simple solution that can be used (it is easy to replace logstash with your own database/logging solution):

First install the dependencies (the first only needs to be installed on the machine that will keep the logs, the second needs to be installed on all workers):

```
brew install logstash
pip install python-logstash
```

Start the logstash server on one of the machines (here localhost):

```
logstash -e 'output { stdout {} }'
```

(make sure to wait till it comes up, it takes 1-2 minutes).

On the driver then run (replace localhost with the ip of the machine running the logstash server):

```
import logging
import logstash
import ray

def setup_logger(worker):
    logger = logging.getLogger('logger')
    logger.setLevel(logging.INFO)
    logger.addHandler(logstash.TCPLogstashHandler('localhost', 5000, version=1))
    # Assigning ray.logger is sketchy and shouldn't be done. This should be maintained as part
    # of the application logic.
    ray.logger = logger

# Note that the following function is not part of the public API at the moment
ray.worker.global_worker.run_function_on_all_workers(setup_logger)
ray.init()

ray.logger.info("This is from the driver")

@ray.remote
def f():
    ray.logger.info("This is from a worker")

ray.get(f.remote())

@ray.remote
class Actor(object):
     def __init__(self):
        ray.logger.info("And this is from an actor")

a = Actor.remote()
```

**Comment by mitar** [comment]
Yes. This works great. I use something similar to store things in MongoDB.

**Comment by mitar** [comment]
(I also redirect all stdout/stderr to logging first, and then from logging to MongoDB. And then on driver from MongoDB to console.)

**Comment by mitar** [comment]
> `ray.logger = logger` 

Are you sure this is necessary? I thought that Ray now uses per Python module logger and not one global logger?

**Comment by pcmoritz** [comment]
The logger object (which in this case is connected to logstash) needs to be assigned somwhere so it can be accessed from user code. It could be a global variable but then it needs to be brought into scope with `global`, or it an be assigned to a variable in a module.

How are you handling this?

**Comment by pcmoritz** [comment]
Ah, actually we can use `logging.getLogger` to access the logging object I think.

**Comment by mitar** [comment]
What what? OK, I am NOT doing this. :-) I haven't noticed this in your code. A great thing about Python logging is that you have to redirect only the root logger. Messages propagate up to it (if not filtered). So you just attach your handler to the root logger and this is it.

```python
logging_queue = queue_module.Queue(-1)

queue_handler = handlers.QueueHandler(logging_queue)
queue_handler.setLevel('DEBUG')

root = logging.getLogger()
root.addHandler(queue_handler)
```

BTW, you should add to logger first a memory queue. And then you connect to that memory queue and you redirect it to an external service. In this way logging call will never block when you have connectivity issues.

So I have then also:

```python
handler = utils.MongoDBHandler(mongo_client)
listener = handlers.QueueListener(logging_queue, handler)
listener.start()

atexit.register(listener.stop)
```

BTW, I have @robertnishihara and @richardliaw access to our repo with pointers to our logging code. If you want to look into it and maybe see if you can port something to Ray (to store things to Redis for example) or to write documentation, I could give it to you as well. I do not have time at the moment to make a PR myself to Ray.

**Comment by pcmoritz** [comment]
My code only shows how to set up the logging on all the workers and actors, which logging mechanism is used is an implementation detail :) I.e. you could use an async logger, or your queue implementation or logstash, that's orthogonal.

**Comment by mitar** [comment]
I am just saying that you should not example the code which uses directly logstash, but you should have a queue in between (except if they don't already do queuing in memory themselves).

**Comment by mitar** [comment]
Read more information here: https://docs.python.org/3/howto/logging-cookbook.html#dealing-with-handlers-that-block

**Comment by mitar** [comment]
So there are two important points in my snippet(s):
* Adding handler to the root logger which captures all logging, so you do not need to even know which exactly logger is being used for logging. It propagates up to the root (except if filtered).
* Using a queue handler between real handler and logger. Prevents logger from blocking ever.

**Comment by ericl** [comment]
Auto-closing stale issue.

**Comment by mitar** [comment]
Hm, there was no warning that this will be closed? Generally, one first announce that an issue is stale. And close it later if nobody objects.
